File: Prácticas/P5/NapakalakiGame/lib/CultistPlayer.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'Treasure'
require_relative 'Monster'
require_relative 'Treasurekind'
require_relative 'CombatResult'
require_relative 'Dice'
require_relative 'Napakalaki'
require_relative 'Player'

module NapakalakiGame
class CultistPlayer < Player  
  attr_accessor :myCultistCard    
  #Atributo de Clase
  @@totalCultistPlayers = 0
  def initialize(player, cultist)
    super(player.name)
    copyConstructor(player)
    @myCultistCard = cultist
    @@totalCultistPlayers += 1 #Incrementamos el contador de cultistPlayers
  end  
  #Métodos
  def getCombatLevel
    #En ruby super es una llamada al metodo padre, asi que lo asignamos
    #a una variable y le sumamos el specialValue
    cl = super      
    cl += @myCultistCard.getSpecialValue
    return cl
  end
  def shouldConvert
    #Siempre devuelve false
    return false
  end
  def getOponentLevel(monster) 
    return monster.getSpecialValue
  end
  def computeGoldCoinsValue(treasures)      
    #Devolvemos el valor
    return super.computeGoldCoinsValue(treasures) * 2
  end
  #Metodo de la clase
  def self.totalCultistPlayers
    return @@totalCultistPlayers
  end        
end
end

File: Prácticas/P5/Napakalaki/src/NapakalakiGame/CultistPlayer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author Mario Rodríguez
 */
public class CultistPlayer extends Player {
    
    private static int totalCultistPlayers = 0 ;
    private Cultist myCultistCard ;
    
    /*  Llamará al nuevo constructor de copia
        de Player. Asimismo, incrementará en uno la variable de clase
        totalCultistPlayers. */
    public CultistPlayer( Player p, Cultist c ) 
    {
        super(p) ;
        this.myCultistCard = c ;
        totalCultistPlayers++ ;
    }
    
    /*  Para calcular su nivel de combate sumando el getCombatLevel() de Player 
        más lo que devuelve el método getSpecialValue() de su carta de sectario (Cultist). */
    @Override
    protected int getCombatLevel()
    {
        return super.getCombatLevel() + this.myCultistCard.getSpecialValue() ;
    }
    
    @Override
    protected boolean shouldConvert()
    {
        return false ;
    }
    
    @Override
    protected int getOponentLevel( Monster m )
    {
        return m.getSpecialValue() ;
    }
    
    // Se redefine de manera que las piezas de oro dupliquen su valor a la hora de canjearlas por niveles.
    @Override
    protected float computeGoldCoinsValue( ArrayList< Treasure > t )
    {
        return super.computeGoldCoinsValue(t) * 2 ;
    }
    
    public static int getTotalCultistPlayers()
    {
        return totalCultistPlayers;
    }
    
    public Cultist getCultistCard()
    {
        return this.myCultistCard ;
    }   
}
